man/man1/bplgt.1
.TH bplgt 1 "December 16, 2024" "Logtalk 3.86.0" "Logtalk Documentation"

.SH NAME
bplgt \- Logtalk integration script for B-Prolog

.SH SYNOPSIS
.B bplgt
[\fB-g \fIgoal\fR]

.SH DESCRIPTION
\fBbplgt\fR is a Logtalk integration script for B-Prolog. It automatically loads the B-Prolog adapter file, the Logtalk compiler and runtime, the Logtalk default table of library names and paths, and any existing settings file found on the lookup directories. It uses the \fBbp\fR executable \fB-i\fR option to load the Logtalk files. It passes any options to the \fBbp\fR executable; see the B-Prolog documentation for the available options.

.SH OPTIONS
The most common option is:
.TP
.B \-g
.I goal
Initialization goal, written between double quotes, to be executed after Logtalk startup.

.SH REQUIREMENTS
B-Prolog 8.1 or a later version is required. This integration script looks for a \fBbp\fR executable in the system path.

.SH ENVIRONMENT
.TP
.B LOGTALKHOME
Specifies the location of the Logtalk installation.
.TP
.B LOGTALKUSER
Specifies the location of the Logtalk user directory.

.SH FILES
.TP
.BI $LOGTALKHOME/integration/logtalk_bp.pl
Prolog file used for loading the necessary files for starting Logtalk using B-Prolog as the backend compiler.
.TP
.BI $LOGTALKHOME/adapters/b.pl
The B-Prolog adapter file.
.TP
.BI $LOGTALKHOME/core/core.pl
The Logtalk core compiler and runtime file.
.TP
.BI $LOGTALKHOME/paths/paths.pl
The default table of library names and paths.
.TP
.BI ./settings.lgt
Logtalk settings file (for overriding default settings in the adapter files). It may not exist.
.TP
.BI $LOGTALKUSER/settings.lgt
Logtalk settings file (for overriding default settings in the adapter files). It may not exist.
.TP
.BI $HOME/settings.lgt
Logtalk settings file (for overriding default settings in the adapter files). It may not exist.
.TP
.BI $XDG_CONFIG_HOME/logtalk/settings.lgt
Logtalk settings file (for overriding default settings in the adapter files). It may not exist.
.TP
.BI $HOME/.config/logtalk/settings.lgt
Logtalk settings file (for overriding default settings in the adapter files). It may not exist.

.SH "SEE ALSO"
The Logtalk website at https://logtalk.org/
.PP
The Logtalk User and Reference Manuals at \fB$LOGTALKHOME/manuals/index.html\fR
.PP
The B-Prolog website at http://www.probp.com/
.PP
logtalk_user_setup(1),\ logtalk_version_select(1),\ logtalk_backend_select(1)

.SH AUTHOR
Paulo Moura <pmoura@logtalk.org>

.SH COPYRIGHT
Copyright (c) 1998-2024 Paulo Moura.


RELEASE_NOTES.md
________________________________________________________________________

This file is part of Logtalk <https://logtalk.org/>  
SPDX-FileCopyrightText: 1998-2024 Paulo Moura <pmoura@logtalk.org>  
SPDX-License-Identifier: Apache-2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
________________________________________________________________________


RELEASE NOTES
=============


3.87.0 - January ??, 2025
=========================

Logtalk compiler and runtime
----------------------------

* ADDED: New `includes/2`, `object/3`, `protocol/3`, and `category/3`
properties to the `logtalk::loaded_file_property/2` predicate.

Installers and installation scripts
-----------------------------------

* UPDATED: Synced the MacPorts portfile with the version used upstream.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* FIXED: VSCode support for finding predicate implementations when the
definition is found in an included file.

* FIXED: VSCode support for finding predicate references when the call is
found in an included file.

* FIXED: VSCode support for finding predicate declarations when the reference
is found in an included file.

* FIXED: VSCode support for navigating call hierarchies where an exception
would be thrown in some cases.


3.86.0 - December 18, 2024
==========================

Logtalk compiler and runtime
----------------------------

* ADDED: Experimental `logtalk::message_prefix_file/6` hook predicate to
the message printing mechanism.

* FIXED: False left-recursion warning when compiling a grammar rule for
a multifile non-terminal that calls a local non-terminal with the same
indicator.

Prolog adapter and integration files
------------------------------------

* UPDATED: Ciao Prolog adapter file to pretend that modules are supported
to allow loading libraries and examples that require file `use_module/1`
directives.

Documentation
-------------

* ADDED: Notes on encrypting applications using XVM.

* FIXED: Multiple Handbook grammar and spelling typos.

* FIXED: Multiple man pages grammar and spelling typos.

Library
-------

* FIXED: The `os::sleep/1` predicate for XSB to also accept floats.

Tools
-----

* CHANGED: The `lgtunit` tool test compilation warning and error messages to
include the file context for better integration with other developer tools.

* CHANGED: The `diagrams` tool now uses a Unicode glyph for the magnifying
lens instead of a PNG image.

* CHANGED: The `diagrams` tool now uses a Graphviz `box` shape for Prolog
modules instead of a `tab`, which is now only used for directories and
libraries.

* REMOVED: Deprecated `lgt2svg.js` script from the `diagrams` tool.

* ADDED: Experimental `diagrams` tool support for the d2 graph language.

* UPDATED: The `diagrams` tool to support graph languages such as d2 where the
node identifiers are not global.

* UPDATED: The `diagrams` tool to also support the PearAI, Windsurf, PyCharm,
and IDEA text editors when generating diagrams with local file links.

* UPDATED: The `tutor` tool to also explain duplicated predicate/non-terminal
reference linter warnings.

* UPDATED: The `tutor` tool to call the new `logtalk::message_prefix_file/6`
hook predicate to support VSCode displaying warnings and errors augmented with
explanations.

* UPDATED: The `help` tool for the new `logtalk::message_prefix_file/6` hook
predicate.

* FIXED: The `diagrams` tool to also generate an `instance/class` caption for
metaclasses.

* FIXED: Bug in the `diagrams` tool where repeated nodes could be created for
external entities.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would miss
"includes" edges.

* FIXED: Bug in the `diagrams` tool where file loading diagrams would show
duplicated "includes" edges.

* FIXED: The `diagrams` tool support for the graph languages to quote atoms
in predicate labels and predicate node contents when required.

* FIXED: The `diagrams` tool exclude options to allow generating external
nodes for the excluded directories, files, libraries, and entities.

* FIXED: Bug in the `lgtdoc` tool when printing a warning on missing entity
directive `info/1` key.

Examples
--------

* ADDED: New `permutations` example providing benchmarks based on list
permutations. Adapted from Prolog code written and contributed by Paul
Tarau.

* UPDATED: The `assumptions` example with another individual example.

Tests
-----

* ADDED: Additional tests for the left-recursion linter warning.

* ADDED: Additional tests for the Prolog standard `round/1` evaluable.

Installers and installation scripts
-----------------------------------

* IMPROVED: The manual installation script and MacPorts Portfile instructions
for defining the `LOGTALKHOME` and `LOGTALKUSER` environment variables.

IDEs, text editors, and syntax highlighters support
---------------------------------------------------

* UPDATED: VSCode support to use the new `logtalk::message_prefix_file/6`
hook predicate to support displaying warnings and errors augmented with
the `tutor` tool explanations.

* UPDATED: VSCode support to also list the `lgtunit` tool warning and error
messages in the "PROBLEMS" pane.

* FIXED: Compilation errors when loading the VSCode support file using the XSB
backend.


3.85.0 - November 12, 2024
==========================

Logtalk compiler and runtime
----------------------------

* CHANGED: The `logtalk_load_context/2` built-in predicate to type-check
the key argument.


integration/bplgt.sh
#!/usr/bin/env bash

#############################################################################
## 
##   Integration script for B-Prolog
##   Last updated on September 30, 2022
## 
##   This file is part of Logtalk <https://logtalk.org/>  
##   SPDX-FileCopyrightText: 1998-2023 Paulo Moura <pmoura@logtalk.org>
##   SPDX-License-Identifier: Apache-2.0
##   
##   Licensed under the Apache License, Version 2.0 (the "License");
##   you may not use this file except in compliance with the License.
##   You may obtain a copy of the License at
##   
##       http://www.apache.org/licenses/LICENSE-2.0
##   
##   Unless required by applicable law or agreed to in writing, software
##   distributed under the License is distributed on an "AS IS" BASIS,
##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
##   See the License for the specific language governing permissions and
##   limitations under the License.
## 
#############################################################################


if ! [ "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME should be defined first"
	echo "pointing to your Logtalk installation directory!"
	echo "Trying the default locations for the Logtalk installation..."
	if [ -d "/usr/local/share/logtalk" ]; then
		LOGTALKHOME=/usr/local/share/logtalk
		echo "... using Logtalk installation found at /usr/local/share/logtalk"
	elif [ -d "/usr/share/logtalk" ]; then
		LOGTALKHOME=/usr/share/logtalk
		echo "... using Logtalk installation found at /usr/share/logtalk"
	elif [ -d "/opt/local/share/logtalk" ]; then
		LOGTALKHOME=/opt/local/share/logtalk
		echo "... using Logtalk installation found at /opt/local/share/logtalk"
	elif [ -d "/opt/share/logtalk" ]; then
		LOGTALKHOME=/opt/share/logtalk
		echo "... using Logtalk installation found at /opt/share/logtalk"
	elif [ -d "/opt/homebrew/share/logtalk" ]; then
		LOGTALKHOME=/opt/homebrew/share/logtalk
		echo "... using Logtalk installation found at /opt/homebrew/share/logtalk"
	elif [ -d "$HOME/share/logtalk" ]; then
		LOGTALKHOME="$HOME/share/logtalk"
		echo "... using Logtalk installation found at $HOME/share/logtalk"
	elif [ -f "$( cd "$( dirname "$0" )" && pwd )/../core/core.pl" ]; then
		LOGTALKHOME="$( cd "$( dirname "$0" )" && pwd )/.."
		echo "... using Logtalk installation found at $( cd "$( dirname "$0" )" && pwd )/.."
	else
		echo "... unable to locate Logtalk installation directory!" >&2
		echo
		exit 1
	fi
	echo
	export LOGTALKHOME=$LOGTALKHOME
elif ! [ -d "$LOGTALKHOME" ]; then
	echo "The environment variable LOGTALKHOME points to a non-existing directory!" >&2
	echo "Its current value is: $LOGTALKHOME" >&2
	echo "The variable must be set to your Logtalk installation directory!" >&2
	echo
	exit 1
fi

if ! [ "$LOGTALKUSER" ]; then
	echo "The environment variable LOGTALKUSER should be defined first, pointing"
	echo "to your Logtalk user directory!"
	echo "Trying the default location for the Logtalk user directory..."
	echo
	export LOGTALKUSER=$HOME/logtalk
fi

if [ -d "$LOGTALKUSER" ]; then
	if ! [ -f "$LOGTALKUSER/VERSION.txt" ]; then
		echo "Cannot find version information in the Logtalk user directory at $LOGTALKUSER!"
		echo "Creating an up-to-date Logtalk user directory..."
		logtalk_user_setup
	else
		system_version=$(cat "$LOGTALKHOME/VERSION.txt")
		user_version=$(cat "$LOGTALKUSER/VERSION.txt")
		if [ "$user_version" \< "$system_version" ]; then
			echo "Logtalk user directory at $LOGTALKUSER is outdated: "
			echo "    $user_version < $system_version"
			echo "Creating an up-to-date Logtalk user directory..."
			logtalk_user_setup
		fi
	fi
else
	echo "Cannot find \$LOGTALKUSER directory! Creating a new Logtalk user directory"
	echo "by running the \"logtalk_user_setup\" shell script:"
	logtalk_user_setup
fi

LOGTALK_STARTUP_DIRECTORY=$(pwd)
export LOGTALK_STARTUP_DIRECTORY

exec bp -i "$LOGTALKHOME/integration/logtalk_bp.pl" "$@"
